src/Features/ExternalAccess/Copilot/Internal/Completion/CSharpContextProviderService.cs

```diff
@@ -6,80 +6,41 @@
+    : ICSharpCopilotContextProviderService
 {
-    // Exposed for testing
-    public ImmutableArray<IContextProvider> Providers { get; }
+    private readonly ImmutableArray<IContextProvider> _providers = [.. providers];
 
-    [ImportingConstructor]
-    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-    public CSharpContextProviderService([ImportMany] IEnumerable<IContextProvider> providers)
+    public IAsyncEnumerable<IContextItem> GetContextItemsAsync(Document document, int position, IReadOnlyDictionary<string, object> activeExperiments, CancellationToken cancellationToken)
     {
-        Providers = providers.ToImmutableArray();
-    }
-
-    public async IAsyncEnumerable<IContextItem> GetContextItemsAsync(Document document, int position, IReadOnlyDictionary<string, object> activeExperiments, [EnumeratorCancellation] CancellationToken cancellationToken)
-    {
-        var queue = new AsyncQueue<IContextItem>();
-        var tasks = this.Providers.Select(provider => Task.Run(async () =>
-        {
-            try
-            {
-                await provider.ProvideContextItemsAsync(document, position, activeExperiments, ProvideItemsAsync, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception exception) when (FatalError.ReportAndCatchUnlessCanceled(exception, ErrorSeverity.General))
-            {
-            }
-        },
-        cancellationToken));
-
-        // Let all providers run in parallel in the background, so we can steam results as they come in.
-        // Complete the queue when all providers are done.
-        _ = Task.WhenAll(tasks)
-            .ContinueWith((_, __) => queue.Complete(),
-                          null,
-                          cancellationToken,
-                          TaskContinuationOptions.ExecuteSynchronously,
-                          TaskScheduler.Default);
-
-        while (true)
-        {
-            IContextItem item;
-            try
-            {
-                item = await queue.DequeueAsync(cancellationToken).ConfigureAwait(false);
-            }
-            catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
-            {
-                // Dequeue is cancelled because the queue is empty and completed, we can break out of the loop.
-                break;
-            }
-
-            yield return item;
-        }
-
-        ValueTask ProvideItemsAsync(ImmutableArray<IContextItem> items, CancellationToken cancellationToken)
-        {
-            foreach (var item in items)
+        return ProducerConsumer<IContextItem>.RunAsync(
+            static async (callback, args, cancellationToken) =>
             {
-                queue.Enqueue(item);
-            }
-
-            return default;
-        }
+                var (@this, document, position, activeExperiments) = args;
+                await RoslynParallel.ForEachAsync(
```

Comment:
this automatically parallelizes, prefering to utilize all cores to finish a fixed set of work that is then added to as work items complete.  this is better than kicking off N items (which may be far mroe than the number of cores), causing them to contend with each other, and trying to .WhenAll them.